### ISSUE TITLE
Change Request: expose `shouldUseFlatConfig` (or a wrapper API that selects between the two formats automatically)

### ISSUE THREAD

**Issue by cprussin** [eslint/eslint#17129]
### ESLint version

v8.38.0

### What problem do you want to solve?

Currently, neither [`findFlatConfigFile`](https://github.com/eslint/eslint/blob/d85efad655deacc0dc3fdbbace33307094c3b91b/lib/eslint/flat-eslint.js#L266-L271) nor [`shouldUseFlatConfig`](https://github.com/eslint/eslint/blob/d85efad655deacc0dc3fdbbace33307094c3b91b/lib/cli.js#L283-L301) are exported.  This makes it impossible for tools that use the API directly such as [`jest-runner-eslint`](https://github.com/jest-community/jest-runner-eslint/pull/176) to support both config formats without manually reproducing the behavior in those functions.

### What do you think is the correct solution?

Ideally, it would be great if there were an API exposed (probably in [unsupported-apis.js](https://github.com/eslint/eslint/blob/d85efad655deacc0dc3fdbbace33307094c3b91b/lib/unsupported-api.js)) that would wrap the `FlatESLint` and `ESLint` apis and select the right one automatically, behaving the same as the CLI.

Short of that, exposing `shouldUseFlatConfig` or `findFlatConfigFile` is enough to prevent tools having to manually roll their own logic here.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

I'm happy to submit the fix for this, I'll just need guidance by those who would approve it on which, if either, approach would be acceptable.

**Comment by nzakas** [comment]
I'm not clear on the ask here: are you asking for a way to determine if flat config should be used? Or are you asking for a way to find the flat config file?

There is already an open issue for adding a method to `FlatESLint` to find the config file:
https://github.com/eslint/eslint/issues/17046

**Comment by cprussin** [comment]
@nzakas not exactly the same, we don't need to _find_ the flat config file, we need to know which config format is being used so that we can call the correct APIs (and remove options that aren't supported by the `FlatESLint` API but are supported by the `ESLint` api).  See https://github.com/jest-community/jest-runner-eslint/pull/176/files#diff-f5651cdb0e581e985fe919453f1a58dfba958912e3c5e7ddfb4f31057483eb17R106-R166 for the specific use case.

I think the ideal ask is to expose an API that behaves exactly as the CLI does -- meaning it would determine which config format is being used and would call `ESLint` or `FlatESLint` as needed transparently.  Short of that, the ask is to expose [shouldUseFlatConfig](https://github.com/eslint/eslint/blob/d85efad655deacc0dc3fdbbace33307094c3b91b/lib/cli.js#L283-L301) so that at least that function [doesn't have to be copy-pasted](https://github.com/jest-community/jest-runner-eslint/pull/176/files#diff-f5651cdb0e581e985fe919453f1a58dfba958912e3c5e7ddfb4f31057483eb17R106-R123) if a tool wants to support both config formats simultaneously, as the ESLint CLI does.

**Comment by nzakas** [comment]
> I think the ideal ask is to expose an API that behaves exactly as the CLI does

I don't think we can do this and have it make sense. The CLI operates on a couple of different signals. If `eslint.config.js` is present, then it uses flat config, but also if the `ESLINT_USE_FLAT_CONFIG` environment variable is set to `true` ([see the docs](https://eslint.org/docs/latest/use/configure/configuration-files-new)). Eventually, all of this will go away, so I'm also not sure we want to have this written in stone in some API.

With the new `FlatESLint#findConfigFile()` method, you can determine if a `eslint.config.js` file is present, would that be enough?

```js
let eslint = new FlatESLint(options);
if (!(await eslint.findConfigFile())) {
   eslint = new ESLint(options);
}
```

**Comment by cprussin** [comment]
It's an improvement, and it's probably good enough.  But it's still not _perfect_, as ideally tooling like `jest-runner-eslint` that effectively behave as a drop-in replacement for the CLI would behave more or less identically to the CLI when it comes to choosing between flat & legacy config formats (i.e. matching support for the `ESLINT_USE_FLAT_CONFIG` env variable would be nice).

Would you be opposed to exposing this kind of a wrapper API in `unsupported-api.js`?  That way, it's treated as exactly as stable as `FlatESLint` itself, and is explicitly _not_ set in stone.

If that's not feasible I'm sure it's fine to use `findConfigFile` as a reasonable enough proxy :)

**Comment by cprussin** [comment]
@nzakas might be easier to have some code to be more concrete about what I'm proposing, this is super quick & dirty and almost certainly not correct, and obviously I haven't done any of the tests or anything, but this is roughly what I had in mind: https://github.com/eslint/eslint/commit/5560fcae6a94aae4c009b512864e0f7585688d9b.  Thoughts?

**Comment by ljharb** [comment]
@nzakas tbh if the logic is likely to change in the future that's a strong argument *for* exposing this API - that way, userland doesn't have to care what it's doing, only that it gives them a boolean. Currently, changing the rules will break tools that are attempting to mimic the rules - but with this API, changing the rules would Just Work for those tools.

**Comment by fasttime** [comment]
I think the idea is to remove the current `ESLint` class and rename `FlatESLint` to `ESLint` in the next major version. At that point, `shouldUseFlatConfig` would start resolving to `true` all the time.

**Comment by ljharb** [comment]
sure, but it would still be convenient to have the implementation be `return true` and let consumers not have to care about it.

**Comment by nzakas** [comment]
> as ideally tooling like jest-runner-eslint that effectively behave as a drop-in replacement for the CLI would behave more or less identically to the CLI when it comes to choosing between flat & legacy config formats (i.e. matching support for the ESLINT_USE_FLAT_CONFIG env variable would be nice).

@cprussin ah okay, this was the bit I really didn't quite understand. I don't use Jest, so I wasn't aware that this was how `jest-runner-eslint` would work. Given that, I think we can look at including such a `shouldUseFlatConfig()` function in the unsupported API and just flip it to always return true in the future.

@mdjermanovic what do you think?

**Comment by cprussin** [comment]
that sounds good @nzakas

**Comment by cprussin** [comment]
Just to clarify, would you all prefer a `shouldUseFlatConfig` over something like [`getESLintEngine`](https://github.com/eslint/eslint/commit/5560fcae6a94aae4c009b512864e0f7585688d9b#diff-0fe2421d538797c86d2d2875bc04ea9bb1023c4451f58b9c439000bb22a0cd4eR19-R25)?  The reason I ask is that I expect that the main (or possibly only) use case for `shouldUseFlatConfig` would be to build a function that looks exactly like `getESLintEngine` anyways...

**Comment by cprussin** [comment]
@mdjermanovic makes sense, fair enough.

Done in https://github.com/eslint/eslint/pull/17169.

Thanks folks!
